LiveLines.Spotify/SpotifyService.cs

```diff
@@ -0,0 +1,103 @@
+{
+    private static readonly HttpClient HttpClient = new ();
+    
+    private readonly ISpotifyCredentialsStore _spotifyCredentialsStore;
+    private readonly string _spotifyClientId;
+    private readonly string _spotifyClientSecret;
+    
+    public SpotifyService(IConfiguration configuration, ISpotifyCredentialsStore spotifyCredentialsStore)
+    {
+        _spotifyClientId = configuration.GetValue<string>("SPOTIFY_CLIENT_ID");
+        _spotifyClientSecret = configuration.GetValue<string>("SPOTIFY_CLIENT_SECRET");
+        _spotifyCredentialsStore = spotifyCredentialsStore;
+    }
+
+    public async Task<SpotifyCredentials?> GetSpotifyCredentials(LoggedInUser user)
+    {
+        var spotifyCredentials = await _spotifyCredentialsStore.GetCredentialsForUser(user);
+
+        if (spotifyCredentials == null)
+        {
+            return null;
+        }
+        
+        if (spotifyCredentials.ExpiresAt < DateTime.UtcNow)
+        {
+            return await RefreshAccessToken(spotifyCredentials.RefreshToken);
+        }
+
+        return spotifyCredentials;
+    }
+
+    private async Task<SpotifyCredentials> RefreshAccessToken(string refreshTokenRequest)
+    {
+        var data = new Dictionary<string, string>
+        {
+            {"refresh_token", refreshTokenRequest},
+            {"grant_type", "refresh_token"},
+        };
+
+        var credentialsResponse = await GetSpotifyCredentialsFromOptions(data);
+
+        return GetCredentialsFromResponse(credentialsResponse);
+    }
+
+    public async Task UpsertSpotifyCredentials(LoggedInUser user, string code, string redirectUrl)
+    {
+        var data = new Dictionary<string, string>
+        {
+            {"code", code},
+            {"grant_type", "authorization_code"},
+            {"redirect_uri", redirectUrl},
+        };
+
+        var credentialsResponse = await GetSpotifyCredentialsFromOptions(data);
+        var spotifyCredentials = GetCredentialsFromResponse(credentialsResponse);
+
+        await _spotifyCredentialsStore.UpsertCredentialsForUser(user, spotifyCredentials);
+    }
+    
+    private SpotifyCredentials GetCredentialsFromResponse(SpotifyCredentialsResponse credentialsResponse)
+    {
+        var (accessToken, tokenType, scope, expiresIn, refreshToken) = credentialsResponse;
+        
+        // give 5 minutes of leeway
+        var expiresAt = DateTime.UtcNow.AddSeconds(expiresIn - 60 * 5);
+        return new SpotifyCredentials(accessToken, tokenType, scope, expiresAt, refreshToken);
+    }
+
+    private async Task<SpotifyCredentialsResponse> GetSpotifyCredentialsFromOptions(Dictionary<string, string> data)
+    {
+        var request = new HttpRequestMessage(HttpMethod.Post, "https://accounts.spotify.com/api/token");
+        request.Content = new FormUrlEncodedContent(data);
+
+        var authValue = Convert.ToBase64String(System.Text.Encoding.UTF8.GetBytes($"{_spotifyClientId}:{_spotifyClientSecret}"));
```

Comment:
Consider `System.Text.Encoding.UTF8.GetBytes` -> `UTF8.GetBytes` for readability